narrow: Show streams:all notice only after "oldest" is found.

ryanreh99 opened the pull request:
Followup of PR #15273

timabbott commented:
This looks great!  I think the commit message should have explained why we needed to move this from `narrow.js`, which I think I understand, but is sorta a key detail here.

timabbott commented:
Could we move this either inside or directly adjacent ot the call to `notifications.hide_or_show_history_limit_message` from within `message_fetch.js`?  That feels like a more appropriate place, since then we don't need a comment explaining that we've already updated fetch_status`?  And in any case, these belong next to each other as they're very similar features.

timabbott commented:
(Also, let me know how you've tested this, and post a screenshot of CI, or disable CI on your fork, so we have CI output here).

ryanreh99 commented:
Tested this by creating new user and the notice is not present in home view.
Also after displaying the notice on reaching the "oldest" for a search query and then going to the home view, it has the notice hidden.

I added some related refactoring in the second and third commits and updated the commit messages.

![ci](https://user-images.githubusercontent.com/40122794/84591727-50901480-ae5e-11ea-9e22-002d5da8217d.png)

timabbott commented on static/js/narrow.js:
```diff
@@ -1053,4 +1033,58 @@ exports.hide_empty_narrow_message = function () {
+            !filter.is_personal_filter()) {
+            exports.show_end_of_results_notice();
+        }
+    }
```
Should there be a hide call in the else clause here?  It seems like structurally there should be.  I think we can combine the if statements in this block to make it work cleanly; the `const filter` query is cheap.

timabbott commented on static/js/narrow.js:
```diff
@@ -1053,4 +1033,58 @@ exports.hide_empty_narrow_message = function () {
+        exports.hide_history_limit_notice();
+    }
+};
+
```
I think rather than moving things here, the last commit should move these to a new `static/js/message_list_notices.js`; `narrow.js` is already pretty giant, so we shouldn't be moving things that could be a separate module into it.

timabbott commented:
The first two commits were merged as the series ending with 542a3dea776330839c12fdb0b05fcbaf8ab2e560

timabbott commented:
Merged as c7e39ef09081ddbcbeef4c88051c5e24830c145f, after adding a comment at the top of `update_top_of_narrow_notices` to make clear it's intentional that it assumes we hide first.

I did notice a bug we should probably fix while we're thinking about this, though: the href logic in `show_end_of_results_notice` is wrong in the case that one is looking at "Recent topics" or "Settings" or any other overlay that changes the hash fragment at the moment to not just be the current narrow.

We need to compute the URL to suggest from `msg_list.data.filter` and `hash_util.js` logic, I think.

Can you open an issue for this and work on it?  Thanks @ryanreh99!
